### PR TITLE
many: support debian in our CI

### PR DESCRIPTION
**Pull request by morphis**
This adds support for debian based images on Linode to our CI infrastructure.

### CODE REVIEW

File: packaging/ubuntu-16.04/rules [hunk 1]
```diff
@@ -54,7 +54,7 @@ ifeq ($(shell dpkg-vendor --query Vendor),Ubuntu)
     # libudev and udevd so we need to link with it dynamically.
     VENDOR_ARGS=--enable-nvidia-ubuntu --enable-static-libcap --enable-static-libapparmor --enable-static-libseccomp
 else
-    VENDOR_ARGS=--disable-apparmor
+    VENDOR_ARGS=--disable-apparmor --disable-seccomp
```
Comment (zyga) [inline packaging/ubuntu-16.04/rules]: Can you do a second `ifeq` and check for Debian, the else part should be empty.

File: packaging/ubuntu-16.04/snapd.install [hunk 2]
```diff
@@ -17,7 +17,7 @@ data/udev/rules.d/66-snapd-autoimport.rules /lib/udev/rules.d
 data/info /usr/lib/snapd/
 
 # snap-confine stuff
-etc/apparmor.d/usr.lib.snapd.snap-confine.real
+# etc/apparmor.d/usr.lib.snapd.snap-confine.real
```
Comment (zyga) [inline packaging/ubuntu-16.04/snapd.install]: Perhaps we should remove the top-level symlink and just add `packaging/debian-9` and let the CI part add the relevant stuff. Note that for fedora the process is totally different so we need to re-think how this is done.

File: spread.yaml [hunk 3]
```diff
@@ -54,6 +54,9 @@ backends:
                 kernel: Direct Disk
                 image: ubuntu-16.04-64
                 workers: 2
+            - debian-9-64:
+                kernel: GRUB 2
```
Comment (zyga) [inline spread.yaml]: Not sure what kernel this will be for Debian TBH

File: packaging/ubuntu-16.04/rules (hunk 1 above)
Comment (morphis) [inline packaging/ubuntu-16.04/rules]: @zyga You mean a second ifeq within the else part or dropping the else part completely?

File: tests/main/completion/task.yaml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 summary: Check different completions
 
 # ppc64el disabled because of https://github.com/snapcore/snapd/issues/2502
-systems: [-ubuntu-core-16-*, -ubuntu-*-ppc64el]
+systems: [-ubuntu-core-16-*, -ubuntu-*-ppc64el, -debian-*]
```
Comment (zyga) [inline tests/main/completion/task.yaml]: Why is this disabled on Debian?

File: tests/main/install-sideload/task.yaml [hunk 5]
```diff
@@ -1,5 +1,7 @@
 summary: Checks for snap sideload install
 
+systems: [-debian-*]
```
Comment (zyga) [inline tests/main/install-sideload/task.yaml]: Please always indicate why something is disabled on any given system. Otherwise it is just a bug waiting to be discovered.

File: tests/main/completion/task.yaml (hunk 4 above)
Comment (morphis) [inline tests/main/completion/task.yaml]: As commit message says it's somehow failing as the expected output is different. Something I am still looking into.

File: tests/main/install-sideload/task.yaml (hunk 5 above)
Comment (morphis) [inline tests/main/install-sideload/task.yaml]: Will add comments for each.

Comment (zyga) [inline tests/main/install-sideload/task.yaml]: Thanks!

File: tests/main/confinement-classic/task.yaml [hunk 6]
```diff
@@ -4,7 +4,10 @@ details: |
     classic confinement can be executed correctly. There are two variants of
     this test (classic and jailmode) and the snap (this particular one) should
     function correctly in both cases.
-systems: [-ubuntu-core-16-*,-debian-*]
+systems:
+    - -ubuntu-core-16
+    # No confinement (AppArmor, Seccomp) available on these systems
```
Comment (zyga) [inline tests/main/confinement-classic/task.yaml]: But this is classic confinement :) It should just work, except for jailmode. I think we should not disable any such test but instead make the part that exercises confinement dependant on the hidden command that tells us if confinement is enabled or not.

Comment (morphis) [inline tests/main/confinement-classic/task.yaml]: Exactly that is the plan but in a second wave. As soon as all tests are green I want a first version landed. Using a `snap is-forced-devmode` check will then require some refactoring in the tests which I want clearly separate from this to not make the PRs bigger as needed.

This is classic confinement but fails because of jailmode. Lets enable this again in the second wave. Its not like I am throwing a PR and will never care again :-)

**Comment by zyga** [comment]
There are still a few failures in tests, do you want to get a full review of the branch before those go green?

**Comment by morphis** [comment]
@zyga If you want, please shoot your comments. I am still working on the remaining test cases but another review should be fine already at this stage.

File: packaging/ubuntu-16.04/rules (hunk 1 above)
Comment (zyga) [inline packaging/ubuntu-16.04/rules]: A second, chained ifeq so that the else is applied to !ubuntu and !debian

File: tests/main/confinement-classic/task.yaml [hunk 7]
```diff
@@ -5,13 +5,19 @@ details: |
...
+    run_install() {
+        make -C test-snapd-hello-classic clean
+        make -C test-snapd-hello-classic
+        snap install $1 --dangerous ./test-snapd-hello-classic/test-snapd-hello-classic_0.1_*.snap
+    }
+    run_install --classic
+    # FIXME jailmode doesn't work on debian type systems due to missing
+    # confinement yet. This check will be replaced with proper detection
+    # if confinement is supported on the platform we're running.
+    if [ "$SPREAD_SYSTEM" != "debian-*" ]; then
```
Comment (zyga) [inline tests/main/confinement-classic/task.yaml]: I would not expect globbing to work this way? Are you sure?

Comment (zyga) [inline tests/main/confinement-classic/task.yaml]: Per discussion on IRC this is incorrect.

Comment (morphis) [inline tests/main/confinement-classic/task.yaml]: Yeah, thanks for spotting this. One I missed after I found this out myself.

File: tests/main/completion/task.yaml [hunk 8]
```diff
@@ -1,7 +1,9 @@
 summary: Check different completions
 
-# ppc64el disabled because of https://github.com/snapcore/snapd/issues/2502
-systems: [-ubuntu-core-16-*, -ubuntu-*-ppc64el]
+systems:
```
Comment (mvo5) [inline tests/main/completion/task.yaml]: (nitpick) this change looks unrelated. Fine to have it, just point this out. I guess initially debian was disabled here?

File: tests/main/interfaces-dbus/task.yaml [hunk 9]
```diff
@@ -68,6 +68,13 @@ execute: |
     echo "And plug is disconnected by default"
     snap interfaces | MATCH "$DISCONNECTED_PATTERN"
 
+    # FIXME This is a pretty bad check for unsupported confinement
```
Comment (mvo5) [inline tests/main/interfaces-dbus/task.yaml]: I wonder if we should just skip the entire test on debian (avoiding this slightly ugly thing in the middle).

File: tests/upgrade/basic/task.yaml [hunk 10]
```diff
@@ -45,8 +45,11 @@ execute: |
     snap list | grep test-snapd-tools
     test-snapd-tools.echo Hello | grep Hello
     test-snapd-tools.env | grep SNAP_NAME=test-snapd-tools
-    echo Hello > /var/tmp/myevil.txt
-    test-snapd-tools.cat /var/tmp/myevil.txt && exit 1 || true
+
+    if [ -e /sys/kernel/security/apparmor ]; then
```
Comment (mvo5) [inline tests/upgrade/basic/task.yaml]: Maybe a small comment like: `only test if confinement works if we actually have apparmor available` ?

File: tests/main/interfaces-dbus/task.yaml (hunk 9 above)
Comment (morphis) [inline tests/main/interfaces-dbus/task.yaml]: I would like to run the first part of this as it exercise the dbus service setup through the configuration in the snap.yaml. snapd still puts the dbus configuration file in place to allow ownership over bus names which will be still validated if we have snap confinement or not.

File: tests/main/completion/task.yaml (hunk 8 above)
Comment (morphis) [inline tests/main/completion/task.yaml]: Yeah, let me revert that.

Comment (morphis) [inline tests/main/completion/task.yaml]: Ok, I remember my intention for this. The comment above only applies for ppc so I moved this to an expanded yaml array to put the comment directly above the ppc line.

**Comment by zyga** [comment]
I think you need to disable one more test:

```
Expected permission error accessing openvswitch socket with disconnected plug
```

This is on the openvswitch interface test.

**Comment by morphis** [comment]
`This is on the openvswitch interface test.`

Looks like that one was added in between.

File: tests/main/confinement-classic/task.yaml [hunk 11]
```diff
@@ -5,13 +5,19 @@ details: |
...
     function correctly in both cases.
 systems: [-ubuntu-core-16-*]
-environment:
-    INSTALL_FLAGS/classic: --classic
-    INSTALL_FLAGS/classic_and_jailmode: --classic --jailmode
 prepare: |
-    make -C test-snapd-hello-classic clean
-    make -C test-snapd-hello-classic
-    snap install $INSTALL_FLAGS --dangerous ./test-snapd-hello-classic/test-snapd-hello-classic_0.1_*.snap
+    run_install() {
```
Comment (zyga) [inline tests/main/confinement-classic/task.yaml]: I don't get one thing. The `environment` section above spawns two instances of this test, each with its own prepare code. Now you roll this into one where we `run_install --classic` and `run_install --jailmode` in one prepare step. This effectively means that we no longer test `--classic` sans `--jailmode`. Am I missing anything?

File: tests/upgrade/basic/task.yaml [hunk 12]
```diff
@@ -45,8 +45,12 @@ execute: |
     snap list | grep test-snapd-tools
     test-snapd-tools.echo Hello | grep Hello
     test-snapd-tools.env | grep SNAP_NAME=test-snapd-tools
-    echo Hello > /var/tmp/myevil.txt
-    test-snapd-tools.cat /var/tmp/myevil.txt && exit 1 || true
+
+    # only test if confinement works and we actually have apparmor available
+    if [ -e /sys/kernel/security/apparmor ]; then
```
Comment (zyga) [inline tests/upgrade/basic/task.yaml]: Can you please add a note that says this should migrate to the internal command that queries forced devmode.

File: tests/main/confinement-classic/task.yaml (hunk 11 above)
Comment (morphis) [inline tests/main/confinement-classic/task.yaml]: Good spot!

Comment (morphis) [inline tests/main/confinement-classic/task.yaml]: Will merge prepare and excute to work around this.

File: tests/upgrade/basic/task.yaml (hunk 12 above)
Comment (morphis) [inline tests/upgrade/basic/task.yaml]: Done

File: tests/main/confinement-classic/task.yaml (hunk 11 above)
Comment (zyga) [inline tests/main/confinement-classic/task.yaml]: This looks correct now.
